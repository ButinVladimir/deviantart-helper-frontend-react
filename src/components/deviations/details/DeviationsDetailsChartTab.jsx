import React from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import Section from 'react-bulma-components/lib/components/section';
import Container from 'react-bulma-components/lib/components/container';
import {
  Field,
  Label,
  Control,
} from 'react-bulma-components/lib/components/form';
import Button from 'react-bulma-components/lib/components/button';
import Chart from '../shared/DeviationsChartContainer';
import { datetimePickerProps } from '../../../helpers/datepicker-props';

export default function DeviationsDetailsChartTab({
  id,
  title,
  timestampBegin,
  timestampEnd,
  metadata,
  metadataLoading,
  timestampBeginChangeHandler,
  timestampEndChangeHandler,
  submitHandler,
}) {
  const titlesMap = new Map();
  titlesMap.set(id, title);

  const timestampBeginDate = timestampBegin !== null ? new Date(timestampBegin) : null;
  const timestampEndDate = timestampEnd != null ? new Date(timestampEnd) : null;

  return (
    <>
      <Section>
        <Container>
          <Field horizontal>
            <Field.Label>
              <Label>Timestamp range beginning</Label>
            </Field.Label>
            <Field.Body>
              <Field>
                <Control>
                  <DatePicker
                    {...datetimePickerProps}
                    id="timestamp-begin"
                    disabled={metadataLoading}
                    selectsStart
                    name="timestampBegin"
                    startDate={timestampBeginDate}
                    endDate={timestampEndDate}
                    selected={timestampBeginDate}
                    onChange={timestampBeginChangeHandler}
                  />
                </Control>
              </Field>
            </Field.Body>
          </Field>

          <Field horizontal>
            <Field.Label size="normal">
              <Label>Timestamp range end</Label>
            </Field.Label>
            <Field.Body>
              <Field>
                <Control>
                  <DatePicker
                    {...datetimePickerProps}
                    id="timestamp-end"
                    disabled={metadataLoading}
                    selectsEnd
                    name="timestampEnd"
                    startDate={timestampBeginDate}
                    endDate={timestampEndDate}
                    selected={timestampEndDate}
                    onChange={timestampEndChangeHandler}
                  />
                </Control>
              </Field>
            </Field.Body>
          </Field>

          <Field kind="group">
            <Field.Label size="normal" />
            <Field.Body>
              <Field>
                <Control>
                  <Button id="submit-button" color="primary" onClick={submitHandler} loading={metadataLoading}>Submit</Button>
                </Control>
              </Field>
            </Field.Body>
          </Field>
        </Container>
      </Section>

      <Section>
        <Container>
          <Chart
            metadata={metadata || {}}
            titlesMap={titlesMap}
          />
        </Container>
      </Section>
    </>
  );
}

DeviationsDetailsChartTab.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  timestampBegin: PropTypes.number,
  timestampEnd: PropTypes.number,
  // eslint-disable-next-line react/forbid-prop-types
  metadata: PropTypes.object,
  metadataLoading: PropTypes.bool.isRequired,
  timestampBeginChangeHandler: PropTypes.func.isRequired,
  timestampEndChangeHandler: PropTypes.func.isRequired,
  submitHandler: PropTypes.func.isRequired,
};

DeviationsDetailsChartTab.defaultProps = {
  timestampBegin: null,
  timestampEnd: null,
  metadata: null,
};
