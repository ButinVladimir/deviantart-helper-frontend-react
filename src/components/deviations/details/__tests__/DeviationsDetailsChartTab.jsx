import React from 'react';
import { shallow } from 'enzyme';
import DeviationsDetailsChartTab from '../DeviationsDetailsChartTab';

describe('DeviationsDetailsChartTab', () => {
  it('can be rendered correctly when timestamps are null', () => {
    const wrapper = shallow(
      <DeviationsDetailsChartTab
        id="1234-567"
        title="Title"
        timestampBegin={null}
        timestampEnd={null}
        metadata={null}
        metadataLoading={false}
        timestampBeginChangeHandler={() => {}}
        timestampEndChangeHandler={() => {}}
        submitHandler={() => {}}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can be rendered correctly when timestamps are not null', () => {
    const wrapper = shallow(
      <DeviationsDetailsChartTab
        id="1234-567"
        title="Title"
        timestampBegin={2}
        timestampEnd={3}
        metadata={null}
        metadataLoading={false}
        timestampBeginChangeHandler={() => {}}
        timestampEndChangeHandler={() => {}}
        submitHandler={() => {}}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can handle changing timestamp begin', () => {
    const timestampBeginChangeHandler = jest.fn();
    const wrapper = shallow(
      <DeviationsDetailsChartTab
        id="1234-567"
        title="Title"
        timestampBegin={2}
        timestampEnd={3}
        metadata={null}
        metadataLoading={false}
        timestampBeginChangeHandler={timestampBeginChangeHandler}
        timestampEndChangeHandler={() => {}}
        submitHandler={() => {}}
      />,
    );

    wrapper.find('#timestamp-begin').simulate('change');

    expect(timestampBeginChangeHandler).toHaveBeenCalled();
  });

  it('can handle changing timestamp end', () => {
    const timestampEndChangeHandler = jest.fn();
    const wrapper = shallow(
      <DeviationsDetailsChartTab
        id="1234-567"
        title="Title"
        timestampBegin={2}
        timestampEnd={3}
        metadata={null}
        metadataLoading={false}
        timestampBeginChangeHandler={() => {}}
        timestampEndChangeHandler={timestampEndChangeHandler}
        submitHandler={() => {}}
      />,
    );

    wrapper.find('#timestamp-end').simulate('change');

    expect(timestampEndChangeHandler).toHaveBeenCalled();
  });

  it('can handle submitting', () => {
    const submitHandler = jest.fn();
    const wrapper = shallow(
      <DeviationsDetailsChartTab
        id="1234-567"
        title="Title"
        timestampBegin={2}
        timestampEnd={3}
        metadata={null}
        metadataLoading={false}
        timestampBeginChangeHandler={() => {}}
        timestampEndChangeHandler={() => {}}
        submitHandler={submitHandler}
      />,
    );

    wrapper.find('#submit-button').simulate('click');

    expect(submitHandler).toHaveBeenCalled();
  });
});
