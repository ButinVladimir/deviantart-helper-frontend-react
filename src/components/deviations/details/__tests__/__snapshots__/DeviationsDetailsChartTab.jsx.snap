// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviationsDetailsChartTab can be rendered correctly when timestamps are not null 1`] = `
<Fragment>
  <Section
    className=""
    renderAs="section"
    size={null}
    style={Object {}}
  >
    <Container
      breakpoint={null}
      fluid={false}
      renderAs="div"
      style={Object {}}
    >
      <Field
        align={null}
        className=""
        horizontal={true}
        kind={null}
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size={null}
          style={Object {}}
        >
          <Label
            className=""
            htmlFor={null}
            size={null}
            style={Object {}}
          >
            Timestamp range beginning
          </Label>
        </FieldLabel>
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <DatePicker
                allowSameDay={false}
                autoComplete="off"
                className="input"
                dateFormat="dd.MM.yyyy HH:mm"
                dateFormatCalendar="LLLL yyyy"
                disabled={false}
                disabledKeyboardNavigation={false}
                dropdownMode="scroll"
                endDate={1970-01-01T00:00:00.003Z}
                id="timestamp-begin"
                inlineFocusSelectedMonth={false}
                isClearable={true}
                monthsShown={1}
                name="timestampBegin"
                nextMonthButtonLabel="Next month"
                onBlur={[Function]}
                onChange={[Function]}
                onClickOutside={[Function]}
                onFocus={[Function]}
                onInputClick={[Function]}
                onInputError={[Function]}
                onKeyDown={[Function]}
                onMonthChange={[Function]}
                onSelect={[Function]}
                onYearChange={[Function]}
                preventOpenOnFocus={false}
                previousMonthButtonLabel="Previous Month"
                readOnly={false}
                renderDayContents={[Function]}
                selected={1970-01-01T00:00:00.002Z}
                selectsStart={true}
                shouldCloseOnSelect={true}
                showMonthDropdown={true}
                showMonthYearPicker={false}
                showTimeInput={false}
                showTimeSelect={true}
                showYearDropdown={true}
                startDate={1970-01-01T00:00:00.002Z}
                strictParsing={false}
                timeCaption="Time"
                timeFormat="HH:mm"
                timeInputLabel="Time"
                timeIntervals={30}
                withPortal={true}
              />
            </Control>
          </Field>
        </FieldBody>
      </Field>
      <Field
        align={null}
        className=""
        horizontal={true}
        kind={null}
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size="normal"
          style={Object {}}
        >
          <Label
            className=""
            htmlFor={null}
            size={null}
            style={Object {}}
          >
            Timestamp range end
          </Label>
        </FieldLabel>
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <DatePicker
                allowSameDay={false}
                autoComplete="off"
                className="input"
                dateFormat="dd.MM.yyyy HH:mm"
                dateFormatCalendar="LLLL yyyy"
                disabled={false}
                disabledKeyboardNavigation={false}
                dropdownMode="scroll"
                endDate={1970-01-01T00:00:00.003Z}
                id="timestamp-end"
                inlineFocusSelectedMonth={false}
                isClearable={true}
                monthsShown={1}
                name="timestampEnd"
                nextMonthButtonLabel="Next month"
                onBlur={[Function]}
                onChange={[Function]}
                onClickOutside={[Function]}
                onFocus={[Function]}
                onInputClick={[Function]}
                onInputError={[Function]}
                onKeyDown={[Function]}
                onMonthChange={[Function]}
                onSelect={[Function]}
                onYearChange={[Function]}
                preventOpenOnFocus={false}
                previousMonthButtonLabel="Previous Month"
                readOnly={false}
                renderDayContents={[Function]}
                selected={1970-01-01T00:00:00.003Z}
                selectsEnd={true}
                shouldCloseOnSelect={true}
                showMonthDropdown={true}
                showMonthYearPicker={false}
                showTimeInput={false}
                showTimeSelect={true}
                showYearDropdown={true}
                startDate={1970-01-01T00:00:00.002Z}
                strictParsing={false}
                timeCaption="Time"
                timeFormat="HH:mm"
                timeInputLabel="Time"
                timeIntervals={30}
                withPortal={true}
              />
            </Control>
          </Field>
        </FieldBody>
      </Field>
      <Field
        align={null}
        className=""
        horizontal={false}
        kind="group"
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size="normal"
          style={Object {}}
        />
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <Button
                className=""
                color="primary"
                disabled={false}
                fullwidth={false}
                id="submit-button"
                inverted={false}
                isSelected={false}
                isStatic={false}
                loading={false}
                onClick={[Function]}
                outlined={false}
                remove={false}
                renderAs="button"
                reset={false}
                rounded={false}
                size={null}
                state={null}
                style={Object {}}
                submit={false}
                text={false}
              >
                Submit
              </Button>
            </Control>
          </Field>
        </FieldBody>
      </Field>
    </Container>
  </Section>
  <Section
    className=""
    renderAs="section"
    size={null}
    style={Object {}}
  >
    <Container
      breakpoint={null}
      fluid={false}
      renderAs="div"
      style={Object {}}
    >
      <Connect(Chart)
        metadata={Object {}}
        titlesMap={
          Map {
            "1234-567" => "Title",
          }
        }
      />
    </Container>
  </Section>
</Fragment>
`;

exports[`DeviationsDetailsChartTab can be rendered correctly when timestamps are null 1`] = `
<Fragment>
  <Section
    className=""
    renderAs="section"
    size={null}
    style={Object {}}
  >
    <Container
      breakpoint={null}
      fluid={false}
      renderAs="div"
      style={Object {}}
    >
      <Field
        align={null}
        className=""
        horizontal={true}
        kind={null}
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size={null}
          style={Object {}}
        >
          <Label
            className=""
            htmlFor={null}
            size={null}
            style={Object {}}
          >
            Timestamp range beginning
          </Label>
        </FieldLabel>
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <DatePicker
                allowSameDay={false}
                autoComplete="off"
                className="input"
                dateFormat="dd.MM.yyyy HH:mm"
                dateFormatCalendar="LLLL yyyy"
                disabled={false}
                disabledKeyboardNavigation={false}
                dropdownMode="scroll"
                endDate={null}
                id="timestamp-begin"
                inlineFocusSelectedMonth={false}
                isClearable={true}
                monthsShown={1}
                name="timestampBegin"
                nextMonthButtonLabel="Next month"
                onBlur={[Function]}
                onChange={[Function]}
                onClickOutside={[Function]}
                onFocus={[Function]}
                onInputClick={[Function]}
                onInputError={[Function]}
                onKeyDown={[Function]}
                onMonthChange={[Function]}
                onSelect={[Function]}
                onYearChange={[Function]}
                preventOpenOnFocus={false}
                previousMonthButtonLabel="Previous Month"
                readOnly={false}
                renderDayContents={[Function]}
                selected={null}
                selectsStart={true}
                shouldCloseOnSelect={true}
                showMonthDropdown={true}
                showMonthYearPicker={false}
                showTimeInput={false}
                showTimeSelect={true}
                showYearDropdown={true}
                startDate={null}
                strictParsing={false}
                timeCaption="Time"
                timeFormat="HH:mm"
                timeInputLabel="Time"
                timeIntervals={30}
                withPortal={true}
              />
            </Control>
          </Field>
        </FieldBody>
      </Field>
      <Field
        align={null}
        className=""
        horizontal={true}
        kind={null}
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size="normal"
          style={Object {}}
        >
          <Label
            className=""
            htmlFor={null}
            size={null}
            style={Object {}}
          >
            Timestamp range end
          </Label>
        </FieldLabel>
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <DatePicker
                allowSameDay={false}
                autoComplete="off"
                className="input"
                dateFormat="dd.MM.yyyy HH:mm"
                dateFormatCalendar="LLLL yyyy"
                disabled={false}
                disabledKeyboardNavigation={false}
                dropdownMode="scroll"
                endDate={null}
                id="timestamp-end"
                inlineFocusSelectedMonth={false}
                isClearable={true}
                monthsShown={1}
                name="timestampEnd"
                nextMonthButtonLabel="Next month"
                onBlur={[Function]}
                onChange={[Function]}
                onClickOutside={[Function]}
                onFocus={[Function]}
                onInputClick={[Function]}
                onInputError={[Function]}
                onKeyDown={[Function]}
                onMonthChange={[Function]}
                onSelect={[Function]}
                onYearChange={[Function]}
                preventOpenOnFocus={false}
                previousMonthButtonLabel="Previous Month"
                readOnly={false}
                renderDayContents={[Function]}
                selected={null}
                selectsEnd={true}
                shouldCloseOnSelect={true}
                showMonthDropdown={true}
                showMonthYearPicker={false}
                showTimeInput={false}
                showTimeSelect={true}
                showYearDropdown={true}
                startDate={null}
                strictParsing={false}
                timeCaption="Time"
                timeFormat="HH:mm"
                timeInputLabel="Time"
                timeIntervals={30}
                withPortal={true}
              />
            </Control>
          </Field>
        </FieldBody>
      </Field>
      <Field
        align={null}
        className=""
        horizontal={false}
        kind="group"
        multiline={false}
        renderAs="div"
        style={Object {}}
      >
        <FieldLabel
          className=""
          renderAs="div"
          size="normal"
          style={Object {}}
        />
        <FieldBody
          className=""
          renderAs="div"
          style={Object {}}
        >
          <Field
            align={null}
            className=""
            horizontal={false}
            kind={null}
            multiline={false}
            renderAs="div"
            style={Object {}}
          >
            <Control
              className=""
              fullwidth={false}
              iconLeft={false}
              iconRight={false}
              loading={false}
              renderAs="div"
              size={null}
              style={Object {}}
            >
              <Button
                className=""
                color="primary"
                disabled={false}
                fullwidth={false}
                id="submit-button"
                inverted={false}
                isSelected={false}
                isStatic={false}
                loading={false}
                onClick={[Function]}
                outlined={false}
                remove={false}
                renderAs="button"
                reset={false}
                rounded={false}
                size={null}
                state={null}
                style={Object {}}
                submit={false}
                text={false}
              >
                Submit
              </Button>
            </Control>
          </Field>
        </FieldBody>
      </Field>
    </Container>
  </Section>
  <Section
    className=""
    renderAs="section"
    size={null}
    style={Object {}}
  >
    <Container
      breakpoint={null}
      fluid={false}
      renderAs="div"
      style={Object {}}
    >
      <Connect(Chart)
        metadata={Object {}}
        titlesMap={
          Map {
            "1234-567" => "Title",
          }
        }
      />
    </Container>
  </Section>
</Fragment>
`;
